src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,246 @@
+
+                                case ConstantPatternSyntax p:
+                                    return new Pattern.ByValue(p.Expression);
+
+                                default:
+                                    return null;
+                            }
+                        }
+
+                    case SyntaxKind.LogicalAndExpression:
+                        {
+                            // Look for the form "expr && cond" where "expr" must be one of the above cases.
+                            // This will turn into a case guard e.g. "case pat when cond:".
+                            var node = (BinaryExpressionSyntax)operand;
+
+                            // Since "&&" operator has a left associativity, first we need to find the leftmost
+                            // expression in the if-condition, then we can replace it with an appropriate pattern.
+                            var leftmost = GetLeftmostCondition(node);
+
+                            // Make sure that it is one of the expressions that we can handle to prevent recursion.
+                            switch (leftmost.Kind())
+                            {
+                                case SyntaxKind.IsPatternExpression:
+                                case SyntaxKind.IsExpression:
+                                case SyntaxKind.EqualsExpression:
+                                    break;
+
+                                default:
+                                    return null;
+                            }
+
+                            var pattern = CreatePatternFromExpression(leftmost);
+                            if (pattern == null)
+                            {
+                                return null;
+                            }
+
+                            // In case of chained "&&" operators, we need to justify the precedence of operands in which
+                            // the "&&" operator is applied. For example, in "if (expr && cond1 && cond2)" we need to
+                            // separate out the leftmost expression and reconstruct the rest of conditions into a new
+                            // expression, so that they would be evaluated entirely after "expr" e.g. "expr && (cond1 && cond2)".
+                            // Afterwards, we can directly use it in a case guard e.g. "case pat when cond1 && cond2:".
+                            var condition = (leftmost.Parent.Parent as BinaryExpressionSyntax)
+                                ?.WithLeft(((BinaryExpressionSyntax)leftmost.Parent).Right);
+
+                            return new Pattern.Guarded((Pattern)pattern, (condition ?? node.Right).WalkDownParentheses());
+                        }
+
+                    default:
+                        return null;
+                }
+            }
+
+            protected override bool CanConvertIfToSwitch(IfStatementSyntax ifStatement)
```

Comment:
comment why we need this check for C#.  provide a code example to make it clear.  Thanks!